src/StrawberryShake/Client/src/Transport.WebSockets/SocketConnectionPool.cs

```diff
@@ -0,0 +1,126 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace StrawberryShake.Transport.WebSockets
+{
+    internal sealed class SocketClientPool
+        : ISocketClientPool
+    {
+        private readonly SemaphoreSlim _semaphoreSlim = new(1, 1);
+        private readonly Dictionary<string, ClientInfo> _connections = new();
+        private readonly ISocketClientFactory _socketClientFactory;
+
+        private bool _disposed;
+
+        public SocketClientPool(
+            ISocketClientFactory socketClientFactory)
+        {
+            _socketClientFactory = socketClientFactory
+                ?? throw new ArgumentNullException(nameof(socketClientFactory));
+        }
+
+        public async Task<ISocketClient> RentAsync(
```

Comment:
Rent async tries to returna a existing instance of a socket, if not found creates a new client and initializes the connection (and therefor the protocol)